README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# cryptowatchR

<!-- badges: start -->

[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/cryptowatchR)](https://cran.r-project.org/package=cryptowatchR)
[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-green.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable-1)
[![R-CMD-check](https://github.com/lorenzbr/cryptowatchR/workflows/R-CMD-check/badge.svg)](https://github.com/lorenzbr/cryptowatchR/actions)
[![Codecov test
coverage](https://codecov.io/gh/lorenzbr/cryptowatchR/branch/main/graph/badge.svg)](https://codecov.io/gh/lorenzbr/cryptowatchR?branch=main)
<!-- badges: end -->

An API wrapper for Cryptowatch written in R

## Introduction

This R package provides a wrapper for the Cryptowatch API. You can get
prices and other information (volume, trades, order books, bid and ask
prices, live quotes, and more) about cryptocurrencies and crypto
exchanges. Check <https://docs.cryptowat.ch/rest-api> for a detailed
documentation. The API is free of charge and does not require you to
create an API key. For example, you can easily retrieve historical
prices for a large number of cryptocurrencies without setting up an
account. However, for heavy users, this option is included in the
package as well.

## Installation

You can install the released version of cryptowatchR from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("cryptowatchR")
```

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("lorenzbr/cryptowatchR")
```

## Usage

``` r
## Examples

library(cryptowatchR)

# Some settings
exchange <- "kraken"
pair <- "btcusd"

# Daily prices of Bitcoin in USD
df_ohlc_daily <- get_ohlc(pair)

# Hourly prices
df_ohlc_hourly <- get_ohlc(pair, periods = 3600,
                           before = as.numeric(as.POSIXct(Sys.Date())),
                           after = as.numeric(as.POSIXct(Sys.Date() - 5)),
                           exchange, datetime = FALSE)

# Hourly prices using date/datetime variables
df_ohlc_hourly_datetime <- get_ohlc(pair, periods = 3600, 
                                    before = Sys.Date(), 
                                    after = Sys.Date() - 5,
                                    exchange, datetime = TRUE)

# Daily prices using date/datetime variables
df_ohlc_daily_datetime <- get_ohlc(pair, periods = 86400, 
                                   before = "2021-05-12", 
                                   after = "2021-01-01",
                                   exchange, datetime = TRUE)

# Daily prices using POSIX time
df_ohlc_daily_posix <- get_ohlc(pair, periods = 86400, 
                                before = as.numeric(as.POSIXct("2021-05-12 14:00:00 UCT")),
                                after = as.numeric(as.POSIXct("2021-01-01 14:00:00 UCT")), 
                                exchange, datetime = FALSE)

# Current market price
current_price <- get_current_price("btcusd")
current_prices <- get_current_price()

## Get trades
# Most recent trades (default is 50)
trades <- get_trades(pair)
# 100 trades
trades_100 <- get_trades(pair, limit = 100, datetime = FALSE)
# 200 trades (maximum is 1000) since 1589571417 (unix timestamp)
trades_unix <- get_trades(pair, since = 1589571417, limit = 200, datetime = FALSE)
# 1000 trades and datetime is TRUE
trades_datetime <- get_trades(pair, since = "2021-06-01", limit = 1000)

# 24h-hour summary of cryptocurrency pairs
summary <- get_summary("btcusd")
summaries <- get_summary()
summaries_id <- get_summary(keyBy = "id")
summaries_symbols <- get_summary(keyBy = "symbols")

# Orderbook for Bitcoin-USD (bid and ask with Price and Amount)
orderbook <- get_orderbook(pair, exchange = exchange)
orderbook_depth <- get_orderbook(pair, exchange = exchange, depth = 100)
orderbook_span <- get_orderbook(pair, exchange = exchange, span = 0.5)
orderbook_limit <- get_orderbook(pair, exchange = exchange, limit = 1)

# Get liquidity sums in the order book of Bitcoin in USD
liquidity <- get_orderbook_liquidity("btcusd")

# Live quote for 50 Bitcoins
calculator <- get_orderbook_calculator("btcusd", amount = 50)

# Asset information
df_assets <- get_assets()
asset_btc <- get_assets("btc")

# Information on pairs of currencies
df_pairs <- get_pairs()
pair_btcusd <- get_pairs("btcusd")

# Information on crypto exchanges
df_exchanges <- get_exchanges()
exchange_kraken <- get_exchanges("kraken")
```

## Contact

Please submit bug reports and suggestions via
<https://github.com/lorenzbr/cryptowatchR/issues>

## License

This R package is licensed under the GNU General Public License v3.0.

See
[here](https://github.com/lorenzbr/cryptowatchR/blob/main/LICENSE.md)
for further information.


NEWS.md
# cryptowatchR 0.2.0.9000 (dev)

* Minor style formatting and performance improvements
* Add tests for several functions

# cryptowatchR 0.2.0

* Now `get_markets()` works with all routes (`summary`, `summaries`, `price`, `prices`, `trades` and `orderbook`, `orderbook/liquidity` and `orderbook/calculator`).
* New functions which are wrappers around `get_markets()`: `get_ohlc()`, `get_trades()`, `get_current_prices()`, `get_summary()`, `get_orderbook()`, `get_orderbook_liquidity()` and `get_orderbook_calculator()`
* New feature to use your own API key for all functions. See argument `api_key` in functions.
* New feature to show allowance (costs per request, remaining credits). See argument `allowance` in functions.
* `markets()` is now deprecated and will be removed in a future version. Use `get_ohlc()` instead.

# cryptowatchR 0.1.0

* First release published on CRAN


R/get_trades.R
#' Get trades of cryptocurrencies
#'
#' Get data.frame with trades of cryptocurrencies using the REST API of
#' cryptowat.ch. The route is \emph{trades} and returns trades for a specified
#' market. See \url{https://docs.cryptowat.ch/rest-api/markets/trades} for
#' further information.
#'
#' @usage get_trades(pair, since = NULL, limit = NULL, exchange = "kraken",
#'            datetime = TRUE, api_key = NULL, allowance = FALSE)
#' @param pair A character string containing a pair symbol, e.g. \emph{btcusd}
#' (required argument). Run \code{get_pairs()} to find other available pairs.
#' @param since An integer if \emph{datetime} is \code{FALSE} and a character
#' string if it is \code{TRUE}: Limit the response to trades after this date
#' (optional). This can only be used to filter recent trades. Historical trades
#' cannot be retrieved.
#' @param limit An integer: Limit the number of trades (optional). Max: 1000.
#' @param exchange A character string containing the exchange. Default is
#' \emph{kraken}. Run \code{get_exchanges()} to find other available exchanges.
#' @param datetime A logical. \code{TRUE} indicates that datetime type is used.
#' \code{FALSE} indicates \emph{unix timestamp}. Default is \code{TRUE}.
#' @param api_key A character string containing the API key. See
#' \url{https://docs.cryptowat.ch/rest-api/rate-limit} to learn how to create
#' an account and how to generate an API key.
#' @param allowance A logical (default is \code{FALSE}). If \code{TRUE} the
#' function returns a list which includes allowance information, i.e. cost of
#' the request, remaining credits and your account name.
#'
#' @return A data frame containing trades of a given pair of currencies. If
#' allowance is \code{TRUE}, \code{get_trades()} returns a list.
#'
#' @references See \url{https://docs.cryptowat.ch/rest-api} for further information.
#' @seealso \code{\link{get_markets}}, \code{\link{get_assets}},
#' \code{\link{get_exchanges}}, \code{\link{get_pairs}}
#' @examples
#' \dontrun{
#' # Most recent trades (default is 50)
#' trades <- get_trades(pair)
#' # 200 trades (maximum is 1000) since 1589571417 (unix timestamp)
#' trades_unix <- get_trades(pair, since = 1589571417, limit = 200, datetime = FALSE)
#' # 1000 trades and datetime is TRUE
#' trades_datetime <- get_trades(pair, since = "2021-06-01", limit = 1000)
#' }
#'
#' @export
get_trades <- function(pair, since = NULL, limit = NULL, exchange = "kraken",
                       datetime = TRUE, api_key = NULL, allowance = FALSE) {

  if (datetime)
    if (!is.null(since)) since <- as.numeric(as.POSIXct(since))

  params <- list(since = since, limit = limit)

  trades <- get_markets(route = "trades", pair, exchange,
                        params, api_key, allowance)

  if (allowance) {

    allowance_list <- trades[[2]]
    trades <- trades[[1]]

  }

  df_trades <- as.data.frame(trades)

  names(df_trades) <- c("ID", "Timestamp", "Price", "Amount")

  if (datetime)
    df_trades$Timestamp <- lubridate::as_datetime(df_trades$Timestamp)

  if (allowance) {

    output <- list(result = df_trades, allowance = allowance_list)

  } else if (allowance == FALSE) {

    output <- df_trades

  }

  return(output)

}
